### PR TITLE
Verify release URLs using Trusted Publisher information

### PR DESCRIPTION
**Pull request by facutuesca**
This PR builds on top of the work in https://github.com/pypi/warehouse/pull/15891. The goal is to:

- Add a new `verified` boolean field to each `ReleaseURL` that is set to `True` if the URL matches the Trusted Publisher URL
  - For example, a Source URL of `github.com/user/project` should appear as verified if the release was uploaded using a GitHub Actions Trusted Publisher configured with `github.com/user/project` as its identity.
- Verify the URLs of a release when a new file is uploaded
  - [x] TODO: The current implementation only marks a URL as verified during the first file upload of a release (when the release is created). Change this so that subsequent uploads can also mark URLs as verified.

Closes #15891.

### CODE REVIEW

**Comment by di** [comment]
> TODO: The current implementation only marks a URL as verified during the first file upload of a release (when the release is created). Change this so that subsequent uploads can also mark URLs as verified.

I think this is probably fine as-is for now, as it will cover the majority of use cases, but once we support verifying URLs when attestations are uploaded we should revisit this to re-verify URLs when a file is added to a release (and, likely, add any additional URLs that were not present on the first file, if there are any).

**Comment by facutuesca** [comment]
>  once we support verifying URLs when attestations are uploaded

I thought the URL verification would always be via Trusted Publisher, how do the attestations fit here?

**Comment by di** [comment]
In the future we might have attestations that are not related to a publish that could add additional verification, but I think this is an edge case we can consider later. If we want to include updates here to handle verification for subsequent file uploads that would be fine too!

**Comment by facutuesca** [comment]
@di For verifying URLs of existing releases (when the user uploads files that are not the first file of that release), do we want to verify iff the file is successfully uploaded?

**Comment by di** [comment]
I think only a successful upload makes sense. There are many reasons why an upload would fail, I don't think it's necessary to tease apart which ones would still make verification valid and which ones might not.

**Comment by facutuesca** [comment]
I pushed a commit for verifying also URLs of existing releases. Now, if a file is uploaded for an existing release, we update the verified status of the release's URLs if they were not verified but the current file upload did verify them.

**Review by woodruffw — APPROVED** [review]
LGTM, thanks @facutuesca! We should follow this up with the appropriate project/release view changes 🙂

**Review by firefly9966 — APPROVED** [review]

File: warehouse/migrations/versions/26455e3712a2_create_verified_field_for_releaseurl.py [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+Revises: bb6943882aa9
+Create Date: 2024-04-30 18:40:17.149050
+"""
+
+import sqlalchemy as sa
+
+from alembic import op
+
+revision = "26455e3712a2"
+down_revision = "bb6943882aa9"
```
Comment (di) [inline warehouse/migrations/versions/26455e3712a2_create_verified_field_for_releaseurl.py:25]:
```suggestion
Revises: 208d494aac68
Create Date: 2024-04-30 18:40:17.149050
"""

import sqlalchemy as sa

from alembic import op

revision = "26455e3712a2"
down_revision = "208d494aac68"
```

File: warehouse/packaging/models.py [hunk 2]
```diff
@@ -501,6 +501,7 @@ class ReleaseURL(db.Model):
 
     name: Mapped[str] = mapped_column(String(32))
     url: Mapped[str]
+    verified: Mapped[bool] = mapped_column(default=False)
```
[the remark this answers is not in the document]
Comment (di) [reply warehouse/packaging/models.py:504]: Seems likely, @facutuesca can you update the model to align with what we put in the migration?

As an aside, we might want to think about a CI check for this, not the first time...

[the remark this answers is not in the document]
Comment (miketheman) [reply warehouse/packaging/models.py:504]: > As an aside, we might want to think about a CI check for this, not the first time...

Already working on it! 😉

[the remark this answers is not in the document]
Comment (facutuesca) [reply warehouse/packaging/models.py:504]: @di My bad, I used the model and migration from https://github.com/pypi/warehouse/pull/15891, assuming it was consistent. I'll open a PR to fix it

[the remark this answers is not in the document]
Comment (facutuesca) [reply warehouse/packaging/models.py:504]: https://github.com/pypi/warehouse/pull/16484

**Comment by miketheman** [comment]
This changeset appears to have introduced some warnings in the test suite:

```python
tests/unit/forklift/test_legacy.py::TestFileUpload::test_new_release_url_verified[https://github.com/foo-False]
tests/unit/forklift/test_legacy.py::TestFileUpload::test_new_release_url_verified[https://github.com/foo/bar/-True]
tests/unit/forklift/test_legacy.py::TestFileUpload::test_new_release_url_verified[https://github.com/foo/bar-True]
  /opt/warehouse/src/tests/unit/forklift/test_legacy.py:3906: SAWarning: SELECT statement has a cartesian product between FROM element(s) "releases" and FROM element "release_urls".  Apply join condition(s) between each element to resolve.
    db_request.db.query(ReleaseURL).filter(Release.project == project).one()

tests/unit/forklift/test_legacy.py::TestFileUpload::test_new_release_url_verified[https://google.com-False]
  /opt/warehouse/src/tests/unit/forklift/test_legacy.py:3906: SAWarning: SELECT statement has a cartesian product between FROM element(s) "release_urls" and FROM element "releases".  Apply join condition(s) between each element to resolve.
    db_request.db.query(ReleaseURL).filter(Release.project == project).one()

tests/unit/forklift/test_legacy.py::TestFileUpload::test_new_publisher_verifies_existing_release_url
  /opt/warehouse/src/tests/unit/forklift/test_legacy.py:3984: SAWarning: SELECT statement has a cartesian product between FROM element(s) "releases" and FROM element "release_urls".  Apply join condition(s) between each element to resolve.
    db_request.db.query(ReleaseURL).filter(Release.project == project).all()
```

Curious - why would these query not leverage [the relationship](https://github.com/pypi/warehouse/blob/1185eef19866ff464a0db09aed53525bd2b7d46f/warehouse/packaging/models.py#L500)?

https://github.com/pypi/warehouse/blob/8a5ad08a1bf65b769efaa97ad0f49b7739a20857/tests/unit/forklift/test_legacy.py#L3906
https://github.com/pypi/warehouse/blob/8a5ad08a1bf65b769efaa97ad0f49b7739a20857/tests/unit/forklift/test_legacy.py#L3984

**Comment by facutuesca** [comment]
> This changeset appears to have introduced some warnings in the test suite:

Looking into it now

**Comment by facutuesca** [comment]
> This changeset appears to have introduced some warnings in the test suite:

PR open to fix it here: https://github.com/pypi/warehouse/pull/16528

**Comment by twm** [comment]
PR to prevent recurrence here: https://github.com/pypi/warehouse/pull/16529

**Pull request merged**